lib/semantics/symbol.cc

```diff
@@ -56,19 +57,66 @@ void ObjectEntityDetails::set_type(const DeclTypeSpec &type) {
   type_ = &type;
 }
 
+void ObjectEntityDetails::ReplaceType(const DeclTypeSpec &type) {
+  type_ = &type;
+}
+
 void ObjectEntityDetails::set_shape(const ArraySpec &shape) {
   CHECK(shape_.empty());
   for (const auto &shapeSpec : shape) {
-    shape_.emplace_back(shapeSpec.Clone());
+    shape_.push_back(shapeSpec);
   }
 }
 
+bool ObjectEntityDetails::IsDescriptor() const {
+  if (type_ != nullptr) {
+    if (const IntrinsicTypeSpec * typeSpec{type_->AsIntrinsic()}) {
+      if (typeSpec->category() == TypeCategory::Character) {
+        // TODO maybe character lengths won't be in descriptors
+        return true;
+      }
+    } else if (const DerivedTypeSpec * typeSpec{type_->AsDerived()}) {
+      if (isDummy()) {
+        return true;
+      }
+      // Any length type parameter?
+      if (const Scope * scope{typeSpec->scope()}) {
+        if (const Symbol * symbol{scope->symbol()}) {
+          if (const auto *details{symbol->detailsIf<DerivedTypeDetails>()}) {
+            for (const Symbol *param : details->paramDecls()) {
+              if (const auto *details{param->detailsIf<TypeParamDetails>()}) {
+                if (details->attr() == common::TypeParamAttr::Len) {
+                  return true;
+                }
+              }
+            }
+          }
+        }
+      }
+    } else if (type_->category() == DeclTypeSpec::Category::ClassDerived ||
```

Comment:
ClassDerived will go into the previous branch of the if. AsDerived returns non-null in that case.

Reply:
Right, I didn't remove this case when I extended it.